tests/test.py

```diff
@@ -2273,6 +2280,243 @@ def test_clearing(self):
+            keys_dir=self.current_key_store_path())
+
+        self.engine = sa.create_engine(
+            get_engine_connection_string(
+                self.get_connector_connection_string(self.CONNECTOR_PORT_1),
+                DB_NAME),
+            connect_args=get_connect_args(port=self.CONNECTOR_PORT_1))
+
+        # Remember whether we're running in zone mode. We need to know this
+        # to store and retrieve the data correctly.
+        self.zone_mode = zone_mode
+
+    def stop_services(self):
+        """Gracefully stop Acra services being tested."""
+        self.engine.dispose()
+        stop_process(self.acra_connector)
+        stop_process(self.acra_server)
+
+    @contextlib.contextmanager
+    def running_services(self, **kwargs):
+        self.start_services(**kwargs)
+        try:
+            yield
+        finally:
+            self.stop_services()
+
+    def insert_via_connector(self, data):
+        """Encrypt and insert data via Acra Connector."""
+        # Encryption depends on whether we're using zones or not.
+        if self.zone_mode:
+            acra_struct = create_acrastruct(
+                data.encode('ascii'),
+                read_zone_public_key(
+                    self.zone_id,
+                    self.current_key_store_path()),
+                context=self.zone_id.encode('ascii'))
+        else:
+            acra_struct = create_acrastruct(
+                data.encode('ascii'),
+                read_storage_public_key(
+                    self.client_id,
+                    self.current_key_store_path()))
+
+        row_id = get_random_id()
+        self.engine.execute(test_table.insert(), {
+            'id': row_id, 'data': acra_struct, 'raw_data': data,
+        })
+        return row_id
+
+    def select_via_connector(self, row_id):
+        """Select decrypted data via Acra Connector."""
+        # If we're using zones, zone ID should precede the encrypted data.
+        if self.zone_mode:
+            cols = [sa.cast(self.zone_id.encode('ascii'), BYTEA),
+                    test_table.c.data, test_table.c.raw_data]
+        else:
+            cols = [test_table.c.data, test_table.c.raw_data]
+
+        rows = self.engine.execute(
+            sa.select(cols).where(test_table.c.id == row_id))
+        return rows.first()
+
+    def select_directly(self, row_id):
+        """Select raw data directly from database."""
+        rows = self.engine_raw.execute(
+            sa.select([test_table.c.data]).where(test_table.c.id == row_id))
+        return rows.first()
+
+    def current_key_store_path(self):
+        return os.path.join(self.test_dir.name, '.acrakeys')
+
+    def new_key_store_path(self):
+        return os.path.join(self.test_dir.name, '.acrakeys.new')
+
+    def old_key_store_path(self):
+        return os.path.join(self.test_dir.name, '.acrakeys.old')
+
+    # Now we can proceed with the tests...
+
+    def test_migrate_v1_to_v2(self):
```

Comment:
cool very readable test, and nice pattern to use domain/test specific functions which do meaningful action. forgot name of this pattern in tests *(